### PR TITLE
Berkeley DB v6 compatibility fix

### PR DESCRIPTION
**Pull request by netsafe**
Fixes building error looking like this:
CXX wallet/libbitcoin_wallet_a-db.o
wallet/db.cpp: In member function ‘void CDBEnv::EnvShutdown()’:
wallet/db.cpp:46:16: error: call of overloaded ‘DbEnv(int)’ is ambiguous
DbEnv(0).remove(strPath.c_str(), 0);
^
wallet/db.cpp:46:16: note: candidates are:
In file included from wallet/db.h:21:0,
from wallet/db.cpp:6:
/usr/include/db_cxx.h:916:2: note: DbEnv::DbEnv(const DbEnv&)
DbEnv(const DbEnv &);
^
/usr/include/db_cxx.h:518:2: note: DbEnv::DbEnv(DB_ENV)
DbEnv(DB_ENV _dbenv);
^
/usr/include/db_cxx.h:516:2: note: DbEnv::DbEnv(u_int32_t)
DbEnv(u_int32_t flags);
^
Makefile:5780: recipe for target 'wallet/libbitcoin_wallet_a-db.o' failed
make[2]: *_ [wallet/libbitcoin_wallet_a-db.o] Error 1

### CODE REVIEW

**Comment by netsafe** [comment]
I've originally found it for namecoin - all the detalis there https://github.com/namecoin/namecoin-core/pull/101

**Comment by jonasschnelli** [comment]
I think this change makes sense.
BDB upstream documentation for the `DbEnv` constructor: https://docs.oracle.com/cd/E17276_01/html/api_reference/CXX/envcreate.html

**Comment by luke-jr** [comment]
It might actually be a good idea to fail to build against BDB 6... it changed the license to AGPL, and so users will by default be infringing if they just use Bitcoin Core as-is this way (as AGPL requires providing source code to every peer).

That said, I don't really care strongly either way. Perhaps a clear warning somewhere would be better.

**Comment by netsafe** [comment]
@luke-jr and the code *is* open to every peer, actually - the peer is running an open-source Bitcoin core. What exact do you mean by infringing the license? I see no license is broken here

**Comment by luke-jr** [comment]
@netsafe It's not enough that *someone else* is sharing the code; the person running the software would need to do so, or at least provide notice to each peer how to get it. Right now, Bitcoin Core doesn't do even the notice. That being said, skimming over the AGPL again it seems this requirement exists only for modified versions, so it's probably less of an issue than it seems at face value. But IANAL and it seems like not the best idea to have the license transparently change like this.

In any case, I still don't really care strongly either way. :)

**Comment by netsafe** [comment]
@luke-jr I'm not caring that deep too. As far as I understood this situation, AGPL will fire the way we're speaking **only if it's a client-server** architecture. But in a P2P case - which is a case for Bitcoin - all peers not just should have the same source code, but they're explicitly obliged to : or the network will fall apart because it's nodes won't be able to work together in the same way.

**Comment by luke-jr** [comment]
P2P is just "everyone is a client and server to each other". It's stronger with AGPL, not weaker.

There is absolutely no requirement for all nodes to be running the same source code, only that they be using the same consensus logic, which is a subset of the full program (and doesn't touch the BDB-related code at all).

**Comment by laanwj** [comment]
> It might actually be a good idea to fail to build against BDB 6...

Let's not get involved in copyfights here but focus on the technical side - apparently someone wants to fix building with BDB 6, and if the code changes are acceptable (and don't regress BDB 4/5 support), let's just accept that.

**Comment by netsafe** [comment]
@laanwj It does not breaks 4/5 BDB support - test build confirms it too

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ void CDBEnv::EnvShutdown()
     if (ret != 0)
         LogPrintf("CDBEnv::EnvShutdown: Error %d shutting down database environment: %s\n", ret, DbEnv::strerror(ret));
     if (!fMockDb)
-        DbEnv(0).remove(strPath.c_str(), 0);
+        DbEnv((u_int32_t)0).remove(strPath.c_str(), 0);
```
Comment (laanwj) [inline src/wallet/db.cpp]: Would the c++11 keyword `nullptr` work here?

Comment (netsafe) [inline src/wallet/db.cpp]: No, it must be zero *and* explicitly of type u_int32_t : "The flags parameter must be set to 0." in the docs it's stated strictly

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/8626/commits/323a5fe06af43d3922a435e696d6c52acc6fade1

**Comment by netsafe** [comment]
the configuration string that I'm using for Berkeley DB is - in `build_unix` folder : 
>   ../dist/configure --prefix=/usr --enable-dbm --enable-cxx --enable-compat185 --enable-server --enable-o_direct --with-cryptography=yes --enable-sql --enable-sql_compat --enable-sql_codegen --enable-stl --enable-localization

versions 6.2.23, 5.3.28 and 4.8.30 are producing db_cxx.h that contains the same definition

File: src/wallet/db.cpp (hunk 1 above)
Comment (laanwj) [inline src/wallet/db.cpp]: Oh right! I was misreading, you're not casting 0 to a pointer.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/8626/commits/323a5fe06af43d3922a435e696d6c52acc6fade1

**Comment by sipa** [comment]
utACK 323a5fe06af43d3922a435e696d6c52acc6fade1

**Pull request merged**